File: frontend-app/src/app/api/mappers/workoutMapper.ts
import { Exercise, Workout } from "@/types/workout";
import { WorkoutDTO, ExerciseDTO } from "@/types/workoutDTO";

const mapExerciseDTO = (dto: ExerciseDTO): Exercise => ({
  id: dto.id,
  name: dto.name,
  sets: dto.sets,
  reps: dto.reps,
  weight: dto.weight,
  restTimeSec: dto.restTimeSec,
  muscleGroups: dto.muscleGroups,
});

export const mapWorkoutDTO = (dto: WorkoutDTO): Workout => ({
  id: dto.id,
  title: dto.title,
  scheduledAt: new Date(dto.scheduledAt),
  completedAt: dto.completedAt ? new Date(dto.completedAt) : undefined,
  muscleGroup: dto.muscleGroup ?? "",
  exercises: dto.exercises.map(mapExerciseDTO),
  notes: dto.notes,
});

File: backend-app/apiModule/ApiModule/WebApplication1/Api/WorkoutMapper.cs
using ApiModule.Api.Contracts.Workout;
using ApiModule.Domain;

namespace ApiModule.Api;

public static class WorkoutMapper
{
    public static WorkoutDto ToDto(Workout workout)
        => new(
            workout.Id,
            workout.Title,
            workout.ScheduledAt,
            workout.CompletedAt,
            workout.PerceivedLoad,
            workout.MuscleGroups?.ToArray() ?? [],
            workout.Exercises.Select(ToDto).ToList()
        );

    public static ExerciseDto ToDto(Exercise e)
        => new(
            e.Id,
            e.Name,
            e.Sets,
            e.Reps,
            e.RestTimeSec,
            e.Weight
        );

    public static Workout ToDomain(CreateWorkoutDto dto)
        => new()
        {
            Id = Guid.NewGuid(),
            Title = dto.Title ?? string.Empty,
            ScheduledAt = dto.ScheduledAt,
            CompletedAt = null,
            PerceivedLoad = null,
            MuscleGroups = dto.MuscleGroups?.ToList() ?? [],
            Exercises = dto.Exercises.Select(ToDomain).ToList()
        };

    public static Exercise ToDomain(CreateExerciseDto dto)
        => new()
        {
            Id = Guid.NewGuid(),
            Name = dto.Name ?? string.Empty,
            Sets = dto.Sets ?? 0,
            Reps = dto.Reps ?? 0,
            RestTimeSec = dto.RestTimeSec ?? 0,
            Weight = dto.Weight ?? 0m
        };

    public static Exercise ToDomain(ExerciseDto dto)
        => new()
        {
            Id = dto.Id,
            Name = dto.Name ?? string.Empty,
            Sets = dto.Sets ?? 0,
            Reps = dto.Reps ?? 0,
            RestTimeSec = dto.RestTimeSec ?? 0,
            Weight = dto.Weight ?? 0m
        };
}
